src/meta/src/hummock/compaction/picker/ttl_reclaim_compaction_picker.rs

```diff
@@ -12,28 +12,95 @@
+use risingwave_common::constants::hummock::TABLE_OPTION_DUMMY_RETENTION_SECOND;
+use risingwave_common::util::epoch::Epoch;
+use risingwave_hummock_sdk::compaction_group::StateTableId;
+use risingwave_hummock_sdk::key_range::KeyRangeCommon;
 use risingwave_pb::hummock::hummock_version::Levels;
-use risingwave_pb::hummock::InputLevel;
+use risingwave_pb::hummock::{InputLevel, KeyRange, SstableInfo};
 
 use crate::hummock::compaction::CompactionInput;
 use crate::hummock::level_handler::LevelHandler;
 
+const MIN_TTL_EXPIRE_INTERVAL_MS: u64 = 60 * 60 * 1000; // 1h
+
 #[derive(Default)]
 pub struct TtlPickerState {
-    last_select_index: usize,
+    pub last_select_key_range: KeyRange,
+
+    pub key_range_in_round: KeyRange,
+}
+
+impl TtlPickerState {
+    pub fn valid(&self) -> bool {
+        !self.key_range_in_round.left.is_empty() && !self.key_range_in_round.right.is_empty()
+    }
+
+    pub fn init(&mut self, key_range: KeyRange) {
+        self.key_range_in_round = key_range;
+        self.last_select_key_range = KeyRange {
+            left: vec![],
+            right: self.key_range_in_round.left.clone(),
+            right_exclusive: true,
+        };
+    }
+
+    pub fn clear(&mut self) {
+        self.key_range_in_round = KeyRange::default();
+        self.last_select_key_range = KeyRange::default();
+    }
 }
 
 pub struct TtlReclaimCompactionPicker {
     max_ttl_reclaim_bytes: u64,
-    // todo: filter table option
+    table_id_to_ttl: HashMap<u32, u32>,
 }
 
 impl TtlReclaimCompactionPicker {
-    pub fn new(max_ttl_reclaim_bytes: u64) -> Self {
+    pub fn new(
+        max_ttl_reclaim_bytes: u64,
+        table_id_to_options: HashMap<StateTableId, TableOption>,
+    ) -> Self {
+        let table_id_to_ttl: HashMap<u32, u32> = table_id_to_options
+            .iter()
+            .filter(|id_to_option| {
+                let table_option = id_to_option.1;
+                table_option.retention_seconds.is_some()
+            })
+            .map(|id_to_option| (*id_to_option.0, id_to_option.1.retention_seconds.unwrap()))
+            .collect();
+
         Self {
             max_ttl_reclaim_bytes,
+            table_id_to_ttl,
         }
     }
+
+    fn filter(&self, sst: &SstableInfo, current_epoch_time: u64) -> bool {
+        let table_id_in_sst = sst.table_ids.iter().cloned().collect::<HashSet<u32>>();
+        let expire_epoch =
+            Epoch::from_physical_time(current_epoch_time - MIN_TTL_EXPIRE_INTERVAL_MS);
+
+        for table_id in table_id_in_sst {
+            match self.table_id_to_ttl.get(&table_id) {
+                Some(ttl_second_u32) => {
+                    assert!(*ttl_second_u32 != TABLE_OPTION_DUMMY_RETENTION_SECOND);
+                    // default to zero.
+                    let ttl_mill = (*ttl_second_u32 * 1000) as u64;
+                    let min_epoch = expire_epoch.subtract_ms(ttl_mill);
+                    if Epoch(sst.min_epoch) <= min_epoch {
```

Comment:
Would it be better to maintain per table_id `min_epoch`? Otherwise, the compactor may do unneccessary work if the SST contains both ttl and rarely updated non-ttl tables. If the size of metadata is the concern, we can maintain per table_id `min_epoch` only for L6 SSTs. This is an optimization so it is okay to postpone it to future PRs.

Reply:
We may also skip iterating KV entries by maintaining per table_id `max_epoch` when `max_epoch` < `expired epoch`

Reply:
I think it is not necessary, because in the bottommost level, most sst files would only have one table

Reply:
1. An the bottommost level, most sst will contain only one table_id
2. In other pr, I prefer to split ssts by table_id at the bottommost level